File: strategies/08-nr7-breakout/strategy.py
"""NR7 Volatility-Contraction Breakout Strategy.

Thesis: markets alternate between compression (low-range, low-volume consolidation) and
expansion (breakout). The NR7 pattern (Narrowest Range in 7 bars) identifies the extreme
of compression — bar t has the smallest high-minus-low range of the last n_bars. During
consolidation, stop orders accumulate beyond the edges of the tight range; the first move
outside that range triggers those stops en masse, creating a self-reinforcing price move in
the breakout direction.

Based on Crabel (1990) — both the n_bars=7 label and the exit_bars=4 time-based exit are
published defaults from that work. True range is defined here as high - low (the intraday
range only), not the wider ATR definition that also incorporates prior-close gaps. Python
and ThinkorSwim use this identical definition; the README parity section documents the
divergence from ThinkorSwim's built-in TrueRange() function.
"""

DEFAULT_PARAMS = {"n_bars": 7, "exit_bars": 4}


class NR7Breakout:
    """NR7 Volatility-Contraction Breakout strategy.

    Fires an entry signal when bar t has the narrowest high-minus-low range of
    the last n_bars bars (including bar t). A direction filter selects long or
    short: if close[t] is above the bar midpoint, go long; otherwise go short.
    The position is held for exactly exit_bars bars then unconditionally closed.
    No stop-loss parameter is used — the time-based exit follows Crabel's (1990)
    original design, which deliberately avoids introducing a tuned stop parameter.

    At most one position is held at a time. If a new NR7 signal fires while a
    position is already open, it is ignored until the current position exits.

    True-range definition: high - low. The ATR variant
    (max(H-L, |H-prevClose|, |L-prevClose|)) is NOT used. Both strategy.py and
    strategy.ts use high - low identically for cross-platform parity.

    Args:
        n_bars: Rolling window for the minimum range comparison. Bar t must have
            the narrowest range of the last n_bars bars (including itself) to
            qualify. Default 7 is the Crabel (1990) published value.
        exit_bars: Number of bars to hold the position before unconditional exit.
            Default 4 is the Crabel (1990) published value.

    Raises:
        ValueError: If n_bars < 2 or exit_bars < 1.
    """

    def __init__(self, n_bars: int = 7, exit_bars: int = 4):
        if n_bars < 2:
            raise ValueError("n_bars must be at least 2")
        if exit_bars < 1:
            raise ValueError("exit_bars must be at least 1")
        self.n_bars = n_bars
        self.exit_bars = exit_bars
        self._in_position = False
        self._bars_held = 0
        self._position_direction = 0

    def __call__(self, view) -> float:
        """Compute position signal for the current bar.

        When in a position the hold counter is incremented each bar; once
        exit_bars consecutive holds have been signalled, the position is closed
        and 0.0 is returned. New NR7 signals are ignored while in a position.

        When flat, bar t is tested for the NR7 condition: true range must equal
        the minimum of the last n_bars true ranges (including bar t itself).
        Entry direction is determined by close[t] versus the bar midpoint;
        close above midpoint → long (1.0), at or below → short (-1.0).

        No look-ahead: high[t] and low[t] are fully known at bar t's close.
        The fill executes at bar t+1 open in the engine.

        Args:
            view: A guarded DataFrame slice with OHLCV columns up to and
                including bar t.

        Returns:
            1.0 (long), -1.0 (short), or 0.0 (flat).
        """
        highs = view["high"]
        lows = view["low"]
        closes = view["close"]
        n = len(closes)

        if n < self.n_bars:
            return 0.0

        if self._in_position:
            self._bars_held += 1
            if self._bars_held < self.exit_bars:
                return float(self._position_direction)
            self._in_position = False
            self._bars_held = 0
            self._position_direction = 0
            return 0.0

        true_range = highs - lows
        window_min = float(true_range.iloc[-self.n_bars:].min())
        current_tr = float(true_range.iloc[-1])

        if current_tr != window_min:
            return 0.0

        midpoint = (float(highs.iloc[-1]) + float(lows.iloc[-1])) / 2.0
        current_close = float(closes.iloc[-1])
        direction = 1.0 if current_close > midpoint else -1.0

        self._in_position = True
        self._bars_held = 0
        self._position_direction = direction
        return direction

File: strategies/08-nr7-breakout/strategy.ts
# NR7 Volatility-Contraction Breakout Strategy (Crabel 1990)
# Thesis: bar with the narrowest range in 7 days marks peak volatility compression;
#   accumulated stop-orders beyond the range edges trigger a self-reinforcing breakout.
# Parameters: nBars=7, exitBars=4 (both Crabel 1990 published defaults)
# True-range definition: high - low. This script uses high - low directly rather than
#   the built-in TrueRange() function. ThinkorSwim's TrueRange() adds prior-close gaps:
#   TrueRange() = max(high - low, abs(high - close[1]), abs(low - close[1])).
#   Python strategy.py also uses high - low, so both implementations are identical.
#   On gap-open days the two definitions diverge; high - low is the more conservative
#   measure and matches Crabel's original bar-range definition exactly.
# Parity note: Python engine runs with allow_short=False (default), so short signals
#   are treated as flat. ThinkorSwim implements both long and short sides as written.
# Signal logic: addOrder BUY_AUTO on NR7 long; SELL_TO_CLOSE after exitBars bars
# Load: Charts > Studies > Edit Studies > Strategies tab > Create > paste > apply Daily.
# Fill note: open[-1] = next-bar open, matching Python engine t+1 open fill model.

input nBars = 7;
input exitBars = 4;

# True range = high - low (NOT TrueRange() — see header comment)
def tr = high - low;
def rollingMinTR = Lowest(tr, nBars);
def isNR7 = tr == rollingMinTR;
def midpoint = (high + low) / 2;

# Direction filter: close above midpoint → long candidate; at or below → short candidate
def nr7Long = isNR7 and close > midpoint;
def nr7Short = isNR7 and close <= midpoint;

# Time-based exit counter.
# -1 = no position. 0 = entry bar. 1..exitBars-2 = hold bars. Transitions to -1 at exitBars-1.
# New NR7 while in position is ignored (hold condition takes priority via ordering of branches).
def longCounter;
longCounter = if longCounter[1] >= 0 and longCounter[1] < exitBars - 1
                  then longCounter[1] + 1
              else if nr7Long and longCounter[1] < 0 and shortCounter[1] < 0
                  then 0
              else -1;

def shortCounter;
shortCounter = if shortCounter[1] >= 0 and shortCounter[1] < exitBars - 1
                   then shortCounter[1] + 1
               else if nr7Short and longCounter[1] < 0 and shortCounter[1] < 0
                   then 0
               else -1;

# Entry fires on the bar the counter is set to 0 (transitioning from -1)
def enterLong  = longCounter  == 0 and longCounter[1]  < 0;
def enterShort = shortCounter == 0 and shortCounter[1] < 0;

# Exit fires on the bar the counter drops back to -1 (after holding exitBars bars)
def exitLong  = longCounter  == -1 and longCounter[1]  >= 0;
def exitShort = shortCounter == -1 and shortCounter[1] >= 0;

AddOrder(OrderType.BUY_AUTO,     enterLong,  open[-1], 1, Color.GREEN, Color.GREEN, "NR7 Long Entry");
AddOrder(OrderType.SELL_TO_CLOSE, exitLong,  open[-1], 1, Color.RED,   Color.RED,   "NR7 Long Exit");
AddOrder(OrderType.SELL_SHORT,   enterShort, open[-1], 1, Color.RED,   Color.RED,   "NR7 Short Entry");
AddOrder(OrderType.BUY_TO_CLOSE,  exitShort, open[-1], 1, Color.GREEN, Color.GREEN, "NR7 Short Exit");

# Visualization: highlight NR7 bars with an arrow at the low
plot nr7Mark = if isNR7 then low else Double.NaN;
nr7Mark.SetDefaultColor(Color.YELLOW);
nr7Mark.SetPaintingStrategy(PaintingStrategy.ARROW_UP);

plot trLine = tr;
trLine.SetDefaultColor(Color.GRAY);
trLine.SetLineWeight(1);
